app/src/main/java/com/example/habitsmasher/ui/history/HabitEventItemAdapter.java

```diff
@@ -0,0 +1,95 @@
+package com.example.habitsmasher.ui.history;
+
+import android.content.Context;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.ImageView;
+import android.widget.TextView;
+
+import androidx.annotation.NonNull;
+import androidx.constraintlayout.widget.ConstraintLayout;
+import androidx.fragment.app.FragmentActivity;
+import androidx.recyclerview.widget.RecyclerView;
+
+import com.example.habitsmasher.HabitEvent;
+import com.example.habitsmasher.HabitEventList;
+import com.example.habitsmasher.R;
+import com.firebase.ui.firestore.FirestoreRecyclerAdapter;
+import com.firebase.ui.firestore.FirestoreRecyclerOptions;
+import com.firebase.ui.firestore.ObservableSnapshotArray;
+
+import java.text.DateFormat;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+/**
+ * The HabitEventItemAdapter class
+ * Based on HabitItemAdapter, extension of FirestoreRecycler that shows habit events list with live updates
+ */
+public class HabitEventItemAdapter extends FirestoreRecyclerAdapter<HabitEvent, HabitEventItemAdapter.HabitEventViewHolder> {
+    // Initialize variables
+    private Context _context;
+    private final String DATE_FORMAT = "dd/MM/yyyy";
+    private final ObservableSnapshotArray<HabitEvent> _snapshots;
```

Comment:
Since we now have **two similar adapters**, would it make sense to introduce an **interface** that is shared between the two that extracts the commonalities between the two (context, snapshots, date format, etc.)

Reply:
Hmmm I see where you are coming from, but I think there's actually too many differences between them to support that at this time. Only context and date format are the same, snapshots actually have a different type between the two (HabitEvent vs Habit). I definitely think it's worth looking into in the future, but it may need further discussion so perhaps in a later issue?

Reply:
Yup, I was never implying that we need to do it **here and now**. Just starting up the discussion and asking some clarifying questions. I think there are definitely some similarities that can be extracted away, we basically have two very similar adapters. The type differences can be solved easily using Java Generics, and **interfaces** don't provide implementations, just the declarations of what needs to be there.